h/auth/policy.py

```diff
@@ -7,14 +7,28 @@
 
 @interface.implementer(interfaces.IAuthenticationPolicy)
 class AuthenticationPolicy(object):
-    def __init__(self, api_policy, fallback_policy):
+    def __init__(self, api_policy, fallback_policy, migration_policy=None):
         self.api_policy = api_policy
         self.fallback_policy = fallback_policy
+        self.migration_policy = migration_policy
 
     def authenticated_userid(self, request):
         if _is_api_request(request):
             return self.api_policy.authenticated_userid(request)
-        return self.fallback_policy.authenticated_userid(request)
+
+        userid = self.fallback_policy.authenticated_userid(request)
+        if userid is not None:
+            return userid
+
+        if self.migration_policy is None:
```

Comment:
As before, just add some comments here to explain what's going on for somebody who doesn't have the context in their heads.